### PR TITLE
[FLINK-34456][configuration]Move all checkpoint-related options into CheckpointingOptions

### PR DESCRIPTION
**Pull request by spoon-lz**
## What is the purpose of the change

According to the FLIP-406,Move all checkpoint-related options that are out of `CheckpointingOptions` into `CheckpointingOptions`. Deprecate the original ones.

## Brief change log

- Move all configurations under `org.apache.flink.streaming.api.environment.ExecutionCheckpointingOptions` into `org.apache.flink.configuration.CheckpointingOptions`
- Mark `ExecutionCheckpointingOptions` as deprecated
- Move `org.apache.flink.streaming.api.CheckpointingMode` to  `org.apache.flink.configuration.CheckpointingMode`
- Split `ExternalizedCheckpointCleanup` out of `CheckpointConfig` and move it to flink-core module

## Verifying this change

This change is a minor rework and is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), **Checkpointing**, Kubernetes/Yarn, ZooKeeper: (yes / no / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by spoon-lz** [comment]
@flinkbot run azure

**Comment by spoon-lz** [comment]
@flinkbot  run azure

**Comment by spoon-lz** [comment]
@flinkbot run azure

**Comment by spoon-lz** [comment]
@flinkbot run azure

**Review by Zakelly — COMMENTED** [review]
Hi @spoon-lz , thanks for your contribution! At first glance of your PR, I would suggest keep the old `org.apache.flink.streaming.api.CheckpointingMode` and mark as deprecated. However it is required to start a discussion about such deprecation on dev mailing list first. I will start the discussion and work on the deprecation in Flink 1.19 (will soon be released this month).

**Comment by spoon-lz** [comment]
> Hi @spoon-lz , thanks for your contribution! At first glance of your PR, I would suggest keep the old `org.apache.flink.streaming.api.CheckpointingMode` and mark as deprecated. However it is required to start a discussion about such deprecation on dev mailing list first. I will start the discussion and work on the deprecation in Flink 1.19 (will soon be released this month).

If `CheckpointingMode` is not moved, do we need to copy the same class in the flink-core module? Otherwise, dependency problems will occur.
Or wait until the discussion is over before modifying this PR.

**Comment by Zakelly** [comment]
> > Hi @spoon-lz , thanks for your contribution! At first glance of your PR, I would suggest keep the old `org.apache.flink.streaming.api.CheckpointingMode` and mark as deprecated. However it is required to start a discussion about such deprecation on dev mailing list first. I will start the discussion and work on the deprecation in Flink 1.19 (will soon be released this month).
> 
> If `CheckpointingMode` is not moved, do we need to copy the same class in the flink-core module? Otherwise, dependency problems will occur. Or wait until the discussion is over before modifying this PR.

I suggest both keep the old one and the new introduced one. All the internal usage will use the new one, while only user-facing APIs leverage the old one.

We could wait the conclusion of the discussion.

**Comment by Zakelly** [comment]
@spoon-lz Sorry for the late reply, seems we come to an agreement in discussion. To be brief, we agree to deprecate the old class and introduce a new one. Corresponding APIs will be changed accordingly.
There is one PR ready for this #24381 . I suggest building your PR on top of this after it is merged. WDYT?

**Comment by spoon-lz** [comment]
@Zakelly I also saw the conclusion of the discussion in the email. I will re-edit this PR after #24381 is merged. Another question is, "Split ExternalizedCheckpointCleanup out of CheckpointConfig and move it to flink-core module". Does this change also need to be discussed?

**Comment by Zakelly** [comment]
> @Zakelly I also saw the conclusion of the discussion in the email. I will re-edit this PR after #24381 is merged. Another question is, "Split ExternalizedCheckpointCleanup out of CheckpointConfig and move it to flink-core module". Does this change also need to be discussed?

I think this is much easier since it is only annotated with @\PublicEvolving. I think we can do this in a similar way without another discussion. A seperated PR is also good.

**Comment by spoon-lz** [comment]
@Zakelly A separate PR seems more reasonable, I will submit a separate PR to split `ExternalizedCheckpointCleanup`

**Comment by spoon-lz** [comment]
@Zakelly  There are some references to `org.apache.flink.streaming.api.CheckpointingMode` in `ExecutionCheckpointingOptions`. When we move these configurations into `CheckpointingOptions`, do we need to replace them with `org.apache.flink .core.execution.CheckpointingMode`?

**Comment by Zakelly** [comment]
> @Zakelly There are some references to `org.apache.flink.streaming.api.CheckpointingMode` in `ExecutionCheckpointingOptions`. When we move these configurations into `CheckpointingOptions`, do we need to replace them with `org.apache.flink .core.execution.CheckpointingMode`?

@spoon-lz 
Firstly, we should keep `ExecutionCheckpointingOptions` and mark it as deprecated.
Secondly, any new introduced(moved) options should use the new enum class. Right?

**Comment by spoon-lz** [comment]
@flinkbot run azure

**Comment by spoon-lz** [comment]
@Zakelly New code has been submitted.

**Review by Zakelly — CHANGES_REQUESTED** [review]
@spoon-lz Thanks for the update! There seems to be a conflict. And I think you should re-generate the document as `flink-docs/README.md` says. The documents reference among each other should also be updated if there is any.

File: flink-core/src/main/java/org/apache/flink/configuration/CheckpointingOptions.java
Comment (spoon-lz) [reply flink-core/src/main/java/org/apache/flink/configuration/CheckpointingOptions.java:455]: You're right, I will delete these deprecated options

**Comment by spoon-lz** [comment]
@Zakelly  The code has been modified, but I don’t know why. After I force push to the current branch, I can see the modification in my warehouse, but not in this PR. Do I need to submit a new PR?

**Comment by masteryhx** [comment]
> @Zakelly The code has been modified, but I don’t know why. After I force push to the current branch, I can see the modification in my warehouse, but not in this PR. Do I need to submit a new PR?

Hi, Could you check your branch 'lz-CheckpointingOptions' in your repo ? Seems it doesn't include new commits.

**Comment by spoon-lz** [comment]
@Zakelly @masteryhx Sorry I made the mistake of committing to the wrong branch, the code has been updated.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/ExecutionCheckpointingOptions.java
Comment (masteryhx) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/ExecutionCheckpointingOptions.java:45]: We should also mark it as `@Documentation.ExcludeFromDocumentation` and remove `execution_checkpointing_configuration.html`

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/CheckpointConfig.java
Comment (masteryhx) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/CheckpointConfig.java:165]: Some places in `CheckpointConfig` could also be replaced. e.g.  in `configure` method

**Review by masteryhx — CHANGES_REQUESTED** [review]
Thanks for the PR.
PTAL my comments.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/CheckpointConfig.java
Comment (spoon-lz) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/CheckpointConfig.java:165]: For the methods marked as @Deprecated in `CheckpointConfig`, I have not modified it, just like `ExecutionCheckpointingOptions.CHECKPOINTING_MODE` is still used in CheckpointConfig#getCheckpointingMode.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/ExecutionCheckpointingOptions.java
Comment (spoon-lz) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/ExecutionCheckpointingOptions.java:45]: 1. After adding the annotation, it looks like the `execution checkpointing configuration.html` file needs to be deleted manually.
2. Should all descriptions related to `execution_checkpointing_configuration` in `docs/content/docs/deployment/config.md` be deleted (line 430 to 433),same as `content_hz`.

Comment (masteryhx) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/ExecutionCheckpointingOptions.java:45]: IIUC, all configs will be showed in `checkpointing_configuration.html`, right ?
If true, I think we should remove redundant infos.

Comment (spoon-lz) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/ExecutionCheckpointingOptions.java:45]: Configurations that were added @Deprecated are not moved to `CheckpointingOptions`, and these configurations will not be displayed in `checkpointing_configuration.html`.

Comment (spoon-lz) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/ExecutionCheckpointingOptions.java:45]: @masteryhx Update has been submitted.

**Review by masteryhx — APPROVED** [review]
Thanks for the update.
LGTM.

**Review by Zakelly — APPROVED** [review]
Thanks for the update! LGTM

**Comment by spoon-lz** [comment]
@Zakelly @masteryhx Before this PR was merged, I found a new conflict with `MaterializedTableManager.java` of the master branch. I have resolved this conflict and it can be merged.

**Comment by Zakelly** [comment]
I'll merge this once the CI green.

**Pull request merged**
